main.go

```diff
@@ -0,0 +1,260 @@
+	// with a transaction
+	authenticatedSession, err := env.db.CreateSession(account.AccountId, csrfToken, now)
+	if err != nil {
+		log.Printf("Error creating session: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	err = env.db.DeleteSession(sessionToken)
+	if err != nil {
+		log.Printf("Error deleting old session: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	sessionCookie := http.Cookie{
+		Name:     "session_token",
+		Value:    authenticatedSession.SessionToken,
+		Path:     "/",
+		Secure:   true,
+		HttpOnly: true,
+		SameSite: http.SameSiteLaxMode,
+	}
+
+	http.SetCookie(c.Writer, &sessionCookie)
+	c.AbortWithStatus(http.StatusOK)
+}
+
+func (env *Env) logout(c *gin.Context) {
+	// TODO: move logged in check to its own function
+	now := time.Now()
+	csrfToken := c.Request.Header.Get("CSRF")
+
+	sessionToken, err := c.Cookie("session_token")
+	// should look into deduplicating this
+	// should also look into a logging library w/ levels (debug, info, warning, etc)
+	if err != nil {
+		log.Println("Missing session token")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	session, err := env.db.FetchSession(sessionToken)
+	if err != nil {
+		log.Printf("Error fetching session: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	if session.Expired(now) {
+		log.Println("Session expired")
+		c.AbortWithStatus(http.StatusOK)
+		return
+	}
+
+	if !IsSessionOwner(session, csrfToken) {
+		log.Println("Not owner of session")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	err = env.db.DeleteSession(sessionToken)
+	if err != nil {
+		log.Printf("Error deleting session: %v\n", err)
+	}
+
+	c.AbortWithStatus(http.StatusOK)
+}
+
+func setupRouter(env *Env) *gin.Engine {
+	router := gin.Default()
+	router.LoadHTMLGlob("web/index.html")
+	router.GET("/ping", env.ping)
+	router.GET("/index.html", env.serveIndex)
+	router.GET("/", env.serveIndex)
+	router.POST("/api/login", env.login)
+	router.POST("/api/logout", env.logout)
+
+	router.Static("/static", "./web/static")
+	rootLevelFiles := []string{
+		"asset-manifest.json",
+		"favicon.ico",
+		"logo192.png",
+		"logo512.png",
+		"manifest.json",
+		"robots.txt",
+	}
+
+	for _, file := range rootLevelFiles {
+		router.StaticFile(fmt.Sprintf("/%s", file), fmt.Sprintf("./web/%s", file))
+	}
+
+	return router
+}
+
+func getEnvironment() (*Env, error) {
+	dbUsername, err := os.ReadFile("/run/secrets/mysql_user")
+	if err != nil {
+		return nil, err
+	}
+
+	dbPassword, err := os.ReadFile("/run/secrets/mysql_pw")
+	if err != nil {
+		return nil, err
+	}
+
+	dbName, err := os.ReadFile("/run/secrets/mysql_db")
+	if err != nil {
+		return nil, err
+	}
+
+	db, err := NewMySqlDatabase(
+		strings.TrimSpace(string(dbUsername)),
+		strings.TrimSpace(string(dbPassword)),
+		strings.TrimSpace(string(dbName)),
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &Env{db}, nil
+}
+
+func main() {
```

Comment:
I don't think you want to put everything in your main package (`main.go` and `database.go`). Your `main` package (and file) should basically just parse flags and call some library to do the actual work. This will also make it easier for you to write tests because you can target just that functionality.

For example you may want to do something like below.

```
/pkg/server/server.go
/pkg/server/server_test.go
/pkg/database/database.go
/pkg/database/database_test.go
/cmd
/cmd/server.go
```

Then `/cmd/server.go` which will be your main package would look something like the following:

```go
package main

import (
    "example.com/pkg/server"
)

func main() {
   var hostport := flag.String("hostport", ":8080", "help message for flag")

   server, err := server.New(&server.Config{
      HostPort: *hostport,
   })
   if err != nil {
     log.Fatal(err)
   }
   if err := server.Start(); err != nil {
     log.Fatal(err)
   }
}
```

Reply:
https://github.com/golang-standards/project-layout
I didn't realize that there was a standard(ish) Go project structure. Thanks for introducing it to me!